src/main/java/org/jenkinsci/plugins/plaincredentials/impl/FileCredentialsImpl.java

```diff
@@ -74,34 +119,124 @@
-    public String getData() {
-        return Base64.encodeBase64String(data);
+    /**
+     * Constructor for everyone besides Stapler.
+     *
+     * @param scope       the scope of the credentials.
+     * @param id          the id of the credentials.
+     * @param description the description of the credentials.
+     * @param fileName    the name of the file.
+     * @param secretBytes the content of the file.
+     * @since 1.3
+     */
+    public FileCredentialsImpl(@CheckForNull CredentialsScope scope,
+                               @CheckForNull String id,
+                               @CheckForNull String description, @Nonnull String fileName,
+                               @Nonnull SecretBytes secretBytes) {
+        super(scope, id, description);
+        this.fileName = fileName;
+        this.secretBytes = secretBytes;
     }
 
-    private byte[] unencrypted() throws IOException {
-        try {
-            return KEY.decrypt().doFinal(data);
-        } catch (GeneralSecurityException x) {
-            throw new IOException2(x);
+    /**
+     * Migrate {@link #data} to {@link #secretBytes}
+     *
+     * @return the object.
+     * @throws ObjectStreamException if the data cannot be migrated.
+     */
+    private Object readResolve() throws ObjectStreamException {
+        if (data != null) {
+            // migrate legacy data
+            try {
+                return new FileCredentialsImpl(getScope(), getId(), getDescription(), fileName,
+                        SecretBytes.fromBytes(KEY.decrypt().doFinal(data)));
+            } catch (GeneralSecurityException e1) {
+                InvalidObjectException e2 = new InvalidObjectException(e1.toString());
+                e2.initCause(e1);
+                throw e2;
+            }
         }
+        return this;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public String getFileName() {
+        return fileName;
+    }
+
+    /**
+     * Exposes the encrypted content to jelly.
+     *
+     * @return the encrypted content.
+     */
+    @Restricted(DoNotUse.class) // for Jelly only // TODO consider adding to API
+    public SecretBytes getSecretBytes() {
+        return secretBytes;
     }
 
-    @Override public InputStream getContent() throws IOException {
-        return new ByteArrayInputStream(unencrypted());
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public InputStream getContent() throws IOException {
+        return new ByteArrayInputStream(secretBytes.getPlainData());
     }
 
-    @Extension public static class DescriptorImpl extends BaseStandardCredentialsDescriptor {
+    /**
+     * Our descriptor.
+     */
+    @Extension
+    public static class DescriptorImpl extends BaseStandardCredentialsDescriptor {
 
-        @Override public String getDisplayName() {
+        /**
+         * {@inheritDoc}
+         */
+        @Override
+        public String getDisplayName() {
```

Comment:
All gratuitous. Please avoid making changes unrelated to the stated purpose of the PR.